File: android/app/src/main/java/com/gigiaj/paradise/ShadowDevicePlugin.kt
package com.gigiaj.paradise

import android.content.Context
import android.content.SharedPreferences
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONObject
import org.matrix.rustcomponents.sdk.ClientBuilder
import org.matrix.rustcomponents.sdk.PusherIdentifiers
import org.matrix.rustcomponents.sdk.PusherKind
import org.matrix.rustcomponents.sdk.HttpPusherData
import org.matrix.rustcomponents.sdk.PushFormat
import org.matrix.rustcomponents.sdk.Session
import org.matrix.rustcomponents.sdk.SlidingSyncVersion
import org.matrix.rustcomponents.sdk.SlidingSyncVersionBuilder
import org.matrix.rustcomponents.sdk.Client

@CapacitorPlugin(name = "ShadowDevice")
class ShadowDevicePlugin : Plugin() {

    private suspend fun buildMatrixClient(homeserverUrl: String, ssStr: String?, userId: String): Client {
        val safeUserId = userId.replace(Regex("[^a-zA-Z0-9]"), "_")
        val dataPath = context.filesDir.absolutePath + "/matrix-shadow/$safeUserId/data"
        val cachePath = context.filesDir.absolutePath + "/matrix-shadow/$safeUserId/cache"

        val ssBuilder = when (ssStr) {
            "NATIVE" -> SlidingSyncVersionBuilder.NATIVE
            "NONE" -> SlidingSyncVersionBuilder.NONE
            else -> SlidingSyncVersionBuilder.DISCOVER_NATIVE
        }

        return ClientBuilder()
            .homeserverUrl(homeserverUrl)
            .sessionPaths(dataPath, cachePath)
            .systemIsMemoryConstrained()
            .slidingSyncVersionBuilder(ssBuilder)
            .build()
    }

    private fun getShadowSession(sharedPrefs: SharedPreferences, userId: String): JSONObject? {
        val sessionsStr = sharedPrefs.getString("shadow_sessions", "{}")
        val sessions = JSONObject(sessionsStr)
        return if (sessions.has(userId)) sessions.getJSONObject(userId) else null
    }

    private suspend fun restoreShadowSession(client: Client, sessionData: JSONObject) {
        val accessToken = sessionData.optString("access_token", null)
        val userId = sessionData.optString("user_id", null)
        val deviceId = sessionData.optString("device_id", null)
        val homeserverUrl = sessionData.optString("homeserver_url", null)
        val slidingSyncStr = sessionData.optString("sliding_sync_version", "NONE")

        if (!accessToken.isNullOrEmpty() && !userId.isNullOrEmpty() && !deviceId.isNullOrEmpty() && !homeserverUrl.isNullOrEmpty()) {
            val ssVersion = when (slidingSyncStr) {
                "NATIVE" -> SlidingSyncVersion.NATIVE
                else -> SlidingSyncVersion.NONE
            }

            val session = Session(
                accessToken = accessToken,
                refreshToken = null,
                userId = userId,
                deviceId = deviceId,
                homeserverUrl = homeserverUrl,
                oidcData = null,
                slidingSyncVersion = ssVersion
            )
            client.restoreSession(session)
        } else {
            throw Exception("Missing session credentials in JSON object.")
        }
    }

    @PluginMethod
    fun createSleepyShadow(call: PluginCall) {
        val homeserver = call.getString("homeserverUrl") ?: return call.reject("Missing homeserverUrl")
        val username = call.getString("username") ?: return call.reject("Missing username")
        val password = call.getString("password") ?: return call.reject("Missing password")
        val userId = call.getString("userId") ?: return call.reject("Missing userId")

        val sharedPrefs = context.getSharedPreferences("ParadiseShadow", Context.MODE_PRIVATE)

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val client = buildMatrixClient(homeserver, "DISCOVER", userId)
                client.login(username, password, "Paradise Background Sync", null)
                val session = client.session()
                if (session != null) {
                    val ssName = if (session.slidingSyncVersion == SlidingSyncVersion.NATIVE) "NATIVE" else "NONE"
                    val sessionData = JSONObject().apply {
                        put("access_token", session.accessToken)
                        put("user_id", session.userId)
                        put("device_id", session.deviceId)
                        put("homeserver_url", homeserver)
                        put("sliding_sync_version", ssName)
                    }

                    val sessionsStr = sharedPrefs.getString("shadow_sessions", "{}")
                    val sessions = JSONObject(sessionsStr)
                    sessions.put(session.userId, sessionData)

                    sharedPrefs.edit().putString("shadow_sessions", sessions.toString()).apply()
                }
                val ret = JSObject()
                ret.put("status", "sleepy_shadow_created")
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("Failed to create Sleepy Shadow", e)
            }
        }
    }

    @PluginMethod
    fun activateShadow(call: PluginCall) {
        val fcmToken = call.getString("pushToken") ?: return call.reject("Missing pushToken")
        val pushUrl = call.getString("pushUrl") ?: return call.reject("Missing pushUrl")
        val userId = call.getString("userId") ?: return call.reject("Missing userId")
        val sharedPrefs = context.getSharedPreferences("ParadiseShadow", Context.MODE_PRIVATE)
        val sessionData = getShadowSession(sharedPrefs, userId) ?: return call.reject("No shadow session found for user: $userId")

        val homeserver = sessionData.getString("homeserver_url")
        val ssStr = sessionData.optString("sliding_sync_version", "NONE")

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val client = buildMatrixClient(homeserver, ssStr, userId)
                restoreShadowSession(client, sessionData)
                val identifiers = PusherIdentifiers(
                    pushkey = fcmToken,
                    appId = "moi.paradise.android"
                )

                val httpData = HttpPusherData(
                    url = pushUrl,
                    format = PushFormat.EVENT_ID_ONLY,
                    defaultPayload = null
                )

                client.setPusher(
                    identifiers = identifiers,
                    kind = PusherKind.Http(httpData),
                    appDisplayName = "Paradise",
                    deviceDisplayName = "Android Device",
                    profileTag = "primary",
                    lang = "en"
                )

                sharedPrefs.edit().putString("active_push_user", userId).apply()

                val ret = JSObject()
                ret.put("status", "shadow_activated")
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("Failed to activate Shadow Device", e)
            }
        }
    }

    @PluginMethod
    fun deactivateShadow(call: PluginCall) {
        val fcmToken = call.getString("pushToken") ?: return call.reject("Missing pushToken")
        val userId = call.getString("userId") ?: return call.reject("Missing userId")
        val sharedPrefs = context.getSharedPreferences("ParadiseShadow", Context.MODE_PRIVATE)
        val sessionData = getShadowSession(sharedPrefs, userId) ?: return call.reject("No shadow session found for user: $userId")

        val homeserver = sessionData.getString("homeserver_url")
        val ssStr = sessionData.optString("sliding_sync_version", "NONE")

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val client = buildMatrixClient(homeserver, ssStr, userId)
                restoreShadowSession(client, sessionData)
                val identifiers = PusherIdentifiers(
                    pushkey = fcmToken,
                    appId = "moi.paradise.android"
                )
                client.deletePusher(identifiers)
                sharedPrefs.edit().remove("active_push_user").apply()

                val ret = JSObject()
                ret.put("status", "shadow_deactivated")
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("Failed to deactivate Shadow Device", e)
            }
        }
    }

    @PluginMethod
    fun verifyShadow(call: PluginCall) {
        val userId = call.getString("userId") ?: return call.reject("Missing userId")
        val recoveryKey = call.getString("recoveryKey") ?: return call.reject("Missing recoveryKey")

        val sharedPrefs = context.getSharedPreferences("ParadiseShadow", Context.MODE_PRIVATE)
        val sessionData = getShadowSession(sharedPrefs, userId) ?: return call.reject("No shadow session found for user: $userId")

        val homeserver = sessionData.getString("homeserver_url")
        val ssStr = sessionData.optString("sliding_sync_version", "NONE")

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val client = buildMatrixClient(homeserver, ssStr, userId)
                restoreShadowSession(client, sessionData)
                val encryption = client.encryption()
                encryption.recover(recoveryKey)

                sessionData.put("is_verified", "true")
                val sessionsStr = sharedPrefs.getString("shadow_sessions", "{}")
                val sessions = JSONObject(sessionsStr)
                sessions.put(userId, sessionData)
                sharedPrefs.edit().putString("shadow_sessions", sessions.toString()).apply()

                val ret = JSObject()
                ret.put("status", "shadow_verified")
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("Failed to verify Shadow Device", e)
            }
        }
    }

    @PluginMethod
    fun getShadowStatus(call: PluginCall) {
        val userId = call.getString("userId") ?: return call.reject("Missing userId")
        val sharedPrefs = context.getSharedPreferences("ParadiseShadow", Context.MODE_PRIVATE)
        val sessionData = getShadowSession(sharedPrefs, userId)
        val ret = JSObject()
        if (sessionData != null && sessionData.optString("is_verified") == "true") {
            ret.put("isVerified", true)
        } else {
            ret.put("isVerified", false)
        }
        call.resolve(ret)
    }

    @PluginMethod
    fun getActivePushUser(call: PluginCall) {
        val sharedPrefs = context.getSharedPreferences("ParadiseShadow", Context.MODE_PRIVATE)
        val activeUser = sharedPrefs.getString("active_push_user", null)
        val ret = JSObject()
        if (activeUser != null) {
            ret.put("activeUser", activeUser)
        }
        call.resolve(ret)
    }

    @PluginMethod
    fun setPreviewPreferences(call: PluginCall) {
        val isEnabled = call.getBoolean("enabled", true)
        val sharedPrefs = context.getSharedPreferences("ParadiseShadow", Context.MODE_PRIVATE)
        sharedPrefs.edit().putBoolean("show_previews", isEnabled ?: true).apply()
        call.resolve()
    }
}

File: ios/App/App/ShadowDevicePlugin.swift
import Foundation
import Capacitor
import MatrixRustSDK

@objc(ShadowDevicePlugin)
public class ShadowDevicePlugin: CAPPlugin, CAPBridgedPlugin {
    
    public let identifier = "ShadowDevicePlugin"
    public let jsName = "ShadowDevice"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "createSleepyShadow", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "activateShadow", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deactivateShadow", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "verifyShadow", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getShadowStatus", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getActivePushUser", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setPreviewPreferences", returnType: CAPPluginReturnPromise)
    ]

    let appGroupId = "group.com.gigiaj.paradise"
    
    private func buildMatrixClient(homeserverUrl: String, ssStr: String) async throws -> Client {
        guard let containerUrl = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroupId) else {
            throw NSError(domain: "ShadowDevice", code: 1, userInfo: [NSLocalizedDescriptionKey: "App Group container not found!"])
        }
        
        let dataPath = containerUrl.appendingPathComponent("matrix-shadow/data").path
        let cachePath = containerUrl.appendingPathComponent("matrix-shadow/cache").path
        
        let ssBuilder: SlidingSyncVersionBuilder
        switch ssStr {
        case "NATIVE": ssBuilder = .native
        case "NONE": ssBuilder = .none
        default: ssBuilder = .discoverNative
        }
        
        return try await ClientBuilder()
            .homeserverUrl(url: homeserverUrl)
            .sessionPaths(dataPath: dataPath, cachePath: cachePath)
            .systemIsMemoryConstrained()
            .slidingSyncVersionBuilder(versionBuilder: ssBuilder)
            .build()
    }
    
    private func restoreShadowSession(client: Client, sharedDefaults: UserDefaults) async throws {
        guard let accessToken = sharedDefaults.string(forKey: "access_token"),
              let userId = sharedDefaults.string(forKey: "user_id"),
              let deviceId = sharedDefaults.string(forKey: "device_id"),
              let homeserverUrl = sharedDefaults.string(forKey: "homeserver_url") else {
            throw NSError(domain: "ShadowDevice", code: 2, userInfo: [NSLocalizedDescriptionKey: "Missing session in UserDefaults"])
        }
        
        let slidingSyncStr = sharedDefaults.string(forKey: "sliding_sync_version") ?? "NONE"
        let ssVersion: SlidingSyncVersion = (slidingSyncStr == "NATIVE") ? .native : .none
        
        let session = Session(
            accessToken: accessToken,
            refreshToken: nil,
            userId: userId,
            deviceId: deviceId,
            homeserverUrl: homeserverUrl,
            oidcData: nil,
            slidingSyncVersion: ssVersion
        )
        
        try await client.restoreSession(session: session)
    }


@objc func createSleepyShadow(_ call: CAPPluginCall) {
        let sharedDefaults = UserDefaults(suiteName: appGroupId)!
        
        if call.getString("password") == nil {
            if sharedDefaults.string(forKey: "access_token") != nil {
                call.resolve(["status": "sleepy_shadow_already_exists"])
                return
            } else {
                call.reject("Missing password and no existing shadow session.")
                return
            }
        }
        
        guard let homeserver = call.getString("homeserverUrl"),
              let username = call.getString("username"),
              let password = call.getString("password") else {
            return call.reject("Missing arguments")
        }
        
        sharedDefaults.set(homeserver, forKey: "homeserver_url")
        
        Task {
            do {
                let client = try await buildMatrixClient(homeserverUrl: homeserver, ssStr: "DISCOVER")
                try await client.login(username: username, password: password, initialDeviceName: "Paradise Background Sync", deviceId: nil)
                
                let session = try client.session()
                sharedDefaults.set(session.accessToken, forKey: "access_token")
                sharedDefaults.set(session.userId, forKey: "user_id")
                sharedDefaults.set(session.deviceId, forKey: "device_id")
                
                let ssName = (session.slidingSyncVersion == .native) ? "NATIVE" : "NONE"
                sharedDefaults.set(ssName, forKey: "sliding_sync_version")
                
                call.resolve(["status": "sleepy_shadow_created"])
            } catch {
                call.reject("Failed to create Sleepy Shadow", nil, error)
            }
        }
    }



    @objc func activateShadow(_ call: CAPPluginCall) {
        guard let pushToken = call.getString("pushToken"),
              let pushUrl = call.getString("pushUrl") else {
            return call.reject("Missing arguments")
        }
        
        let sharedDefaults = UserDefaults(suiteName: appGroupId)!
        guard let homeserver = sharedDefaults.string(forKey: "homeserver_url") else {
            return call.reject("No homeserver saved")
        }
        let ssStr = sharedDefaults.string(forKey: "sliding_sync_version") ?? "NONE"


        Task {
            do {
                let client = try await buildMatrixClient(homeserverUrl: homeserver, ssStr: ssStr)
                try await restoreShadowSession(client: client, sharedDefaults: sharedDefaults)
                let payloadDict: [String: Any] = [
                    "aps": [
                        "mutable-content": 1,
                        "alert": [
                            "loc-key": "Notification",
                            "loc-args": []
                        ]
                    ]
                ]
                
                let payloadData = try JSONSerialization.data(withJSONObject: payloadDict)
                guard let payloadString = String(data: payloadData, encoding: .utf8) else {
                    return call.reject("Failed to encode payload string")
                }
                
                let identifiers = PusherIdentifiers(pushkey: pushToken, appId: "moi.paradise.ios")
                let httpData = HttpPusherData(url: pushUrl, format: .eventIdOnly, defaultPayload: payloadString)


                try await client.setPusher(
                    identifiers: identifiers,
                    kind: PusherKind.http(data: httpData),
                    appDisplayName: "Paradise",
                    deviceDisplayName: "iOS Device",
                    profileTag: "primary",
                    lang: "en"
                )
                
                call.resolve(["status": "shadow_activated"])
            } catch {
                call.reject("Failed to activate Shadow Device", nil, error)
            }
        }
    }
    
    
    @objc func deactivateShadow(_ call: CAPPluginCall) {
        guard let pushToken = call.getString("pushToken") else { return call.reject("Missing token") }
        
        let sharedDefaults = UserDefaults(suiteName: appGroupId)!
        guard let homeserver = sharedDefaults.string(forKey: "homeserver_url") else {
            return call.reject("No homeserver saved")
        }
        let ssStr = sharedDefaults.string(forKey: "sliding_sync_version") ?? "NONE"
        
        Task {
            do {
                let client = try await buildMatrixClient(homeserverUrl: homeserver, ssStr: ssStr)
                try await restoreShadowSession(client: client, sharedDefaults: sharedDefaults)
                
                let identifiers = PusherIdentifiers(pushkey: pushToken, appId: "moi.paradise.ios")
                try await client.deletePusher(identifiers: identifiers)
                
                call.resolve(["status": "shadow_deactivated"])
            } catch {
                call.reject("Failed to deactivate Shadow Device", nil, error)
            }
        }
    }
}

